### ISSUE TITLE
No reference found with upsertGraph with existing child and #ref

### ISSUE THREAD

**Issue by johannessjoberg** [Vincit/objection.js#1560]
Hi there!

I was upgrading a service of ours from an objection version of <1.5.0 to 1.6.11. During this one of ours test broke due to `no reference found` errors while upserting with an relation the has a primary key set and uses a #ref for another relation. 

In this particular test setup: 
1. With no primary key: `partOfMovies: [{ name: 'Part of movie updated', movie: { ['#ref']: 'movieUUID' }}, ]` works fine. 
2. With primary key: `partOfMovies: [{ id: 1, name: 'Part of movie updated', movie: { ['#ref']: 'movieUUID' }}, ]` fails.

I think it's easiest if you have a look at the provided reproduction code. First, run it on e.g. 1.4.0 and you should see that it's a success. Then, run it on 1.5.0 or above and you will get the `no reference found error`. From the first looks of it, it seems as if upsertGraph creates a new map of `nodesByUid` when primary key is set, and misses the ones created by it's parent(s).

Big thanks for this great lib and the work you put into it 💟 

```
/**
 * This is a simple template for bug reproductions. It contains three models `Person`, `Animal` and `Movie`.
 * They create a simple IMDB-style database. Try to add minimal modifications to this file to reproduce
 * your bug.
 *
 * install:
 *    npm install objection knex sqlite3 chai
 *
 * run:
 *    node reproduction-template
 */

let Model;

try {
    Model = require('./').Model;
} catch (err) {
    Model = require('objection').Model;
}

const Knex = require('knex');
const chai = require('chai');

async function main() {
    await createSchema();

    ///////////////////////////////////////////////////////////////
    // Your reproduction
    ///////////////////////////////////////////////////////////////

    await PartOfMovie.query().insertGraph({
        id: 1,
        name: 'Part of movie',
    });
    await Person.query().upsertGraph({
        firstName: 'Jennifer',
        lastName: 'Lawrence',

        pets: [
            {
                name: 'Doggo',
                species: 'dog'
            }
        ],

        movies: [
            { 
                ['#id']: 'movieUUID',
                name: 'Movie name'
            }
        ],
        
        partOfMovies: [
            { 
                id: 1, 
                name: 'Part of movie updated', 
                movie: { ['#ref']: 'movieUUID'}
            }, 
        ]
    }, { relate: true, unrelate: true });

    const jennifer = await Person.query()
        .findOne({ firstName: 'Jennifer' })
        .eager('[pets,movies,partOfMovies]');

    console.log('jennifer', jennifer);
    chai.expect(jennifer.pets[0].name).to.equal('Doggo');
    chai.expect(jennifer.partOfMovies[0].movieId).to.equal(jennifer.movies[0].id);
}

///////////////////////////////////////////////////////////////
// Database
///////////////////////////////////////////////////////////////

const knex = Knex({
    client: 'sqlite3',
    useNullAsDefault: true,
    debug: false,
    connection: {
        filename: ':memory:'
    }
});

Model.knex(knex);

///////////////////////////////////////////////////////////////
// Models
///////////////////////////////////////////////////////////////

class Person extends Model {
    static get tableName() {
        return 'Person';
    }

    static get jsonSchema() {
        return {
            type: 'object',
            required: ['firstName', 'lastName'],

            properties: {
                id: { type: 'integer' },
                parentId: { type: ['integer', 'null'] },
                firstName: { type: 'string', minLength: 1, maxLength: 255 },
                lastName: { type: 'string', minLength: 1, maxLength: 255 },
                age: { type: 'number' },

                address: {
                    type: 'object',
                    properties: {
                        street: { type: 'string' },
                        city: { type: 'string' },
                        zipCode: { type: 'string' }
                    }
                }
            }
        };
    }

    static get relationMappings() {
        return {
            pets: {
                relation: Model.HasManyRelation,
                modelClass: Animal,
                join: {
                    from: 'Person.id',
                    to: 'Animal.ownerId'
                }
            },

            movies: {
                relation: Model.ManyToManyRelation,
                modelClass: Movie,
                join: {
                    from: 'Person.id',
                    through: {
                        from: 'Person_Movie.personId',
                        to: 'Person_Movie.movieId'
                    },
                    to: 'Movie.id'
                }
            },

            partOfMovies: {
                relation: Model.ManyToManyRelation,
                modelClass: PartOfMovie,
                join: {
                    from: 'Person.id',
                    through: {
                        from: 'Person_PartOfMovie.personId',
                        to: 'Person_PartOfMovie.movieId'
                    },
                    to: 'PartOfMovie.id'
                }
            },

            children: {
                relation: Model.HasManyRelation,
                modelClass: Person,
                join: {
                    from: 'Person.id',
                    to: 'Person.parentId'
                }
            },

            parent: {
                relation: Model.BelongsToOneRelation,
                modelClass: Person,
                join: {
                    from: 'Person.parentId',
                    to: 'Person.id'
                }
            }
        };
    }
}

class Animal extends Model {
    static get tableName() {
        return 'Animal';
    }

    static get jsonSchema() {
        return {
            type: 'object',
            required: ['name'],

            properties: {
                id: { type: 'integer' },
                ownerId: { type: ['integer', 'null'] },
                name: { type: 'string', minLength: 1, maxLength: 255 },
                species: { type: 'string', minLength: 1, maxLength: 255 }
            }
        };
    }

    static get relationMappings() {
        return {
            owner: {
                relation: Model.BelongsToOneRelation,
                modelClass: Person,
                join: {
                    from: 'Animal.ownerId',
                    to: 'Person.id'
                }
            }
        };
    }
}

class Movie extends Model {
    static get tableName() {
        return 'Movie';
    }

    static get jsonSchema() {
        return {
            type: 'object',
            required: ['name'],

            properties: {
                id: { type: 'integer' },
                name: { type: 'string', minLength: 1, maxLength: 255 }
            }
        };
    }

    static get relationMappings() {
        return {
            actors: {
                relation: Model.ManyToManyRelation,
                modelClass: Person,
                join: {
                    from: 'Movie.id',
                    through: {
                        from: 'Person_Movie.movieId',
                        to: 'Person_Movie.personId'
                    },
                    to: 'Person.id'
                }
            }
        };
    }
}

class PartOfMovie extends Model {
    static get tableName() {
        return 'PartOfMovie';
    }

    static get jsonSchema() {
        return {
            type: 'object',
            required: ['name'],

            properties: {
                id: { type: 'integer' },
                name: { type: 'string', minLength: 1, maxLength: 255 }
            }
        };
    }

    static get relationMappings() {
        return {
            actors: {
                relation: Model.ManyToManyRelation,
                modelClass: Person,
                join: {
                    from: 'PartOfMovie.id',
                    through: {
                        from: 'Person_PartOfMovie.movieId',
                        to: 'Person_PartOfMovie.personId'
                    },
                    to: 'Person.id'
                }
            },
            movie: {
                relation: Model.BelongsToOneRelation,
                modelClass: Movie,
                join: {
                    from: 'PartOfMovie.movieId',
                    to: 'Movie.id'
                }
            }
        };
    }
}

///////////////////////////////////////////////////////////////
// Schema
///////////////////////////////////////////////////////////////

async function createSchema() {
    await knex.schema
        .dropTableIfExists('Person_Movie')
        .dropTableIfExists('Animal')
        .dropTableIfExists('Movie')
        .dropTableIfExists('Person');

    await knex.schema
        .createTable('Person', table => {
            table.increments('id').primary();
            table
                .integer('parentId')
                .unsigned()
                .references('id')
                .inTable('Person');
            table.string('firstName');
            table.string('lastName');
            table.integer('age');
            table.json('address');
        })
        .createTable('Movie', table => {
            table.increments('id').primary();
            table.string('name');
        })
        .createTable('PartOfMovie', table => {
            table.increments('id').primary();
            table.string('name');
            table
                .integer('movieId')
                .unsigned()
                .references('id')
                .inTable('Movie');
        })
        .createTable('Animal', table => {
            table.increments('id').primary();
            table
                .integer('ownerId')
                .unsigned()
                .references('id')
                .inTable('Person');
            table.string('name');
            table.string('species');
        })
        .createTable('Person_Movie', table => {
            table.increments('id').primary();
            table
                .integer('personId')
                .unsigned()
                .references('id')
                .inTable('Person')
                .onDelete('CASCADE');
            table
                .integer('movieId')
                .unsigned()
                .references('id')
                .inTable('Movie')
                .onDelete('CASCADE');
        })
        .createTable('Person_PartOfMovie', table => {
            table.increments('id').primary();
            table
                .integer('personId')
                .unsigned()
                .references('id')
                .inTable('Person')
                .onDelete('CASCADE');
            table
                .integer('movieId')
                .unsigned()
                .references('id')
                .inTable('PartOfMovie')
                .onDelete('CASCADE');
        });
}

main()
    .then(() => {
        console.log('success')
        return knex.destroy()
    })
    .catch(err => {
        console.error(err)
        return knex.destroy()
    });
```

**Comment by koskimas** [comment]
Yep, this is a bug. This happens because you have the `#id` outside and `#ref` inside a subgraph that gets related (PartOrMovie id: 1). I've encountered this before, but thought it was enough of a corner case to be ignored.

This will be pretty difficult to fix. I'd find some other way to implement this.

**Comment by johannessjoberg** [comment]
All right :/ Would it be difficult to implement some kind of injection/sharing of `nodesByUid` to the subgraph?

**Comment by johannessjoberg** [comment]
I'll gladly give it ago :neckbeard: Could you point me in the right direction? Is there something in particular making it difficult? (like concurrency?)

**Comment by koskimas** [comment]
You can't share the nodes as they aren't really nodes in the same graph. You could maybe somehow resolve the references before `upsertGraph` is called recursively in `GraphRecursiveUpsertAction`.  See `_resolveDependencies` method in `GraphInsertAction`. Maybe that could be generalized and moved to `GraphAction` base class?

**Comment by koskimas** [comment]
They could even be resolved already. Maybe it's enough to remove the `#ref` properties from the recursively upserted subgraph? That needs to be done anyway somehow. I don't know if it's ok to mutate the input objects though...

**Comment by koskimas** [comment]
FYI I'm working on a fix for this. It will be released with 2.0

**Comment by koskimas** [comment]
Fixed in 2.0 branch.

**Comment by johannessjoberg** [comment]
Thanks a lot 🙏 I'll look into the code and hopefully learn a thing or two (to be more helpful next time).
